docs/source/notes/extending.rst

```diff
@@ -854,6 +853,62 @@ useful to confirm that a function that isn't present in the dictionary returned
+and return the final result. On the way out, autograd will attach the graph to the
+output and then autocast will have a chance to do any update it needs on exit.
+
+One of the configuration of the dispatcher is the order in which all these feature and backend keys are called. The latest list can be found in ``DispatchKey.h`` inside the ``DispatchKey`` enum. For the purpose of extending torch, the important piece of ordering is: vmap -> Autocast -> Autograd -> ZeroTensor -> Neg/Conj -> Functionalize -> Python -> Backends. The most important key for the purpose of this discussion is the ``Python`` one as every Tensor subclass with the ``__torch_dispatch__`` method defined will have the flag to call into this feature. It is from there that the user-defined method is called and where the behavior can be overwritten arbitrarily. From there, calling the provided ``func`` again will perform a "redispatch".
+
+Some important implication of this implementation are:
+- This code runs "below all features" it is thus only responsible, like a regular backend, to generate the output value of each Tensor (and can, and should, ignore all advanced features like autograd, autocast, etc).
+- If any high level feature implement a give function without redispatching, it will never reach the Python key and so the ``__torch_dispatch__`` callback will never be triggered. This happens in particular for CompositeImplicitAutograd function (composite functions whose autograd formula is specified implicitly by the other native ops it calls) which are evaluated at the Autograd level without redispatching.
+- When calling back to python and when wrapping the results, the usual conversions are used as some objects cannot be represented in python (undefined Tensors for example become None).
+- To have an easy to use python object representing a given native function, they are lazily populated at ``torch.ops.{namespace}.{func_name}.{overload_name}``. The ``func`` object given to ``__torch_dispatch__`` is always an entry from this namespace. This namespace can be used to directly call native ops and bypass the usual python API and binding code.
+
+
+In a similar way where ``__torch_function__`` is able to interpose on all of torch's python API and Tensor methods, ``__torch_dispatch__`` is able to intercept all calls into the aten native API. Note that all methods on Tensors are converted into function calls before entering the dispatcher and thus will appear as function calls here: ``torch.add(a, 2)`` and ``a + 2`` will lead to exactly the same aten call.
+Most of these function are defined in ``native_functions.yaml`` which specifies the properties of these functions as well as their backend implementation. Their registration is then generated via codegen alongside some of the feature registrations.
+Some more exotic functions or features are also registered in other places in the c++ codebase or in user-defined c++ extensions.
```

Comment:
```suggestion
Some more exotic functions or features are also registered in other places in the C++ codebase or in user-defined C++ extensions.
```
This is a cliffhanger--are these exotic functions/features people should be aware of?

Reply:
I think so yes. People reading this should not expect everything to be in native_functions.yaml.